### PR TITLE
Reactive Datasource: support CredentialsProvider changing values

### PR DESCRIPTION
**Pull request by tsegismont**
Instead of:

- creating pgConnectOptions after calling the CredentialsProvider on startup
- using static pool configuration

We:

- use dynamic pool configuration
- call the Credentials provider every time a new connection must be created

This PR allows to support use cases like password rotation.

See also https://issues.redhat.com/projects/QUARKUS/issues/QUARKUS-2995

### CODE REVIEW

**Comment by tsegismont** [comment]
@vietj can you please take a look (dynamic pool config implementation)

@cescoffier @geoand I've made changes so that the credentials provider is invoked on a worker thread, is that fine? The user might be doing blocking calls in the implementation.

**Comment by geoand** [comment]
> I've made changes so that the credentials provider is invoked on a worker thread, is that fine? The user might be doing blocking calls in the implementation

Have you seen such use cases or users mentioning this?

**Comment by tsegismont** [comment]
> Have you seen such use cases or users mentioning this?

I looked into this because of https://issues.redhat.com/projects/QUARKUS/issues/QUARKUS-2995

I haven't tried myself or seen reproducers. I'm trying to think about uses cases, like password rotation. The changing credentials must come from the disk or a remote system? If the lookup is made synchronously in the credentials provider, the event loop will be blocked.

We have two options, I think:

- acknowledging the `CredentialsProvider` contract is synchronous and invoking it on a worker thread
- invoking it on the event loop anyway and then we shall:
  -  document it
  - recommend users to perform the new credentials lookup in a scheduled task

**Comment by geoand** [comment]
I guess we should have it on the worker thread for the time being and if the need arised, look into supporting the `@Blockling` / `@NonBlocking` annotations

**Comment by cescoffier** [comment]
The call to the credential provider must happen on a worker thread unfortunately (wrong design).

**Review by cescoffier — COMMENTED** [review]
The idea looks good.

The credential provider must be called on a worker thread, as it may block.

**Comment by tsegismont** [comment]
> The idea looks good.

Thank you @cescoffier 

> The credential provider must be called on a worker thread, as it may block.

:+1: 



> the only thing i noticed is that the credential manager (vault in my case) seems to be called multiple times on 7 different event loop threads on startup

The crendentials provider can be invoked twice on startup by the extension, and then the default 5 max connections may be created immediately if you setup the database (perhaps with Hibernate Reactive) ?

A good way to check would be to log a fake error (`new Exception()`) in order to see the stack trace.

**Comment by tsegismont** [comment]
@cescoffier @geoand  about credentials provider API: I took a look at the [Vault extension implementation](https://github.com/quarkiverse/quarkus-vault/blob/41adf6d3d8c0c683d6512bacceaf7caf202abfc6/runtime/src/main/java/io/quarkus/vault/runtime/VaultCredentialsProvider.java#L47-L48):

```java
            return vaultDynamicCredentialsManager.getDynamicCredentials(DATABASE_DEFAULT_MOUNT, DEFAULT_REQUEST_PATH,
                    config.databaseCredentialsRole.get()).await().indefinitely();
```

Too bad the contract is synchronous because the implementation uses non blocking-APIs.

Do you think it is worth tracking in a separate issue? In the meantime, we definitely have to invoke this on a worker thread.

**Comment by cescoffier** [comment]
Yes, when I found out about the synchronous and blocking nature of the credential provider it was too late :-(. So, yes, please open an issue to switch to a non-blocking API.

**Comment by tsegismont** [comment]
The PR is ready for review

**Comment by geoand** [comment]
Yeah, totally unrelated

File: extensions/reactive-mssql-client/deployment/src/test/java/io/quarkus/reactive/mssql/client/ChangingCredentialsTestResource.java
Comment (cescoffier) [inline extensions/reactive-mssql-client/deployment/src/test/java/io/quarkus/reactive/mssql/client/ChangingCredentialsTestResource.java:27]: There is a few issues here. `executeAndForget` won't wait for the execution, meaning the the other methods can be called while this one is still in progress. 
I think you need to make sure it's completed before. But now there is a problem as the method is likely going to be invoked on the event loop.

I would recommend switching to an `public void setup(@Observes StartupEvent)` which is called on a worker thread.

**Review by cescoffier — CHANGES_REQUESTED** [review]
Except the postconstruct, the rest looks good to me.

File: extensions/reactive-mssql-client/deployment/src/test/java/io/quarkus/reactive/mssql/client/ChangingCredentialsTestResource.java
Comment (tsegismont) [reply extensions/reactive-mssql-client/deployment/src/test/java/io/quarkus/reactive/mssql/client/ChangingCredentialsTestResource.java:27]: In fact, it is guaranteed that this query will be executed before any other because the pool has a maximum size of 1 connection.

Nevertheless, you are right, using a setup method makes it more clear we want this to be executed before starting the test.

**Review by cescoffier — APPROVED** [review]

**Review by geoand — APPROVED** [review]

**Comment by kdubb** [comment]
@geoand Before this is merged I think there might be a simpler method that I pushed in PR #32204. There is an outstanding issue with that PR that needs to be tracked down but it's a fairly simple change if we can make it work fully.

**Comment by geoand** [comment]
Okay, I'll have a look tomorrow

**Comment by kdubb** [comment]
@tsegismont @geoand My other PR was based on a faulty assumption but the idea was sound. I've opened PR https://github.com/eclipse-vertx/vertx-sql-client/pull/1296 that adds proper support for dynamic credentials to all Vertx SQL clients based on that idea.

That PR supports sourcing dynamic credentials from an external source as well as cycling connections regularly via a maximum connection lifetime.

If we can get that PR accepted would simplify integration into Quarkus to a single line change in each clients pool recorder (e.g. `PgPoolRecorder`) similar to what I had previously implemented but even simpler.

**Comment by kdubb** [comment]
Another point is that, given the current timeline and the amount of time people will continue to use the 2.x series of Quarkus, I'd like to get this fixed for both 2.x series and the upcoming 3.x release. I help maintain the Vault extension and there is an issue filed about this problem at regular intervals in either the extension project or in Quarkus itself.

This is one of the big reasons I am trying to reduce the complexity of the changes in Quarkus; along with the fact that this support really belongs baked into the VertX clients.

**Comment by kdubb** [comment]
The original PR on vertx-sql-client has been split into https://github.com/eclipse-vertx/vertx-sql-client/pull/1298 & https://github.com/eclipse-vertx/vertx-sql-client/pull/1297 at the request of @vietj.

**Comment by tsegismont** [comment]
@kdubb ok let's see what we can do upstream

**Comment by cescoffier** [comment]
What the status on this one? Would it be handled in Vert.x SQL?

**Comment by kdubb** [comment]
@cescoffier Once the two PR's I referenced above are accepted the functionality can be implemented with simple change to the recorder in each reactive sql extension.

They just need to expose and set a `maxLifetime` option and then set the `SqlCredentialsProvider` implementation (which can be shared by all the extensions if there's a common project).

**Comment by kdubb** [comment]
I have had the changes integrated & running for Postgres locally for the past few days. It's working well.

**Comment by tsegismont** [comment]
@cescoffier I've made an intial review of @kdubb 's upstream PR. Please do not merge this PR before we get to a solution upstream.

**Comment by gsmet** [comment]
Marking it as draft until discussions are resolved.

**Comment by tsegismont** [comment]
FTR, I'll have to rework this PR when Vert.x 4.4.2 is released and integrated in Quarkus (see https://github.com/eclipse-vertx/vertx-sql-client/pull/1305)

**Comment by tsegismont** [comment]
I've updated the PR using the new Vert.x API with connect options supplier.

**Comment by kdubb** [comment]
🎉

**Comment by tsegismont** [comment]
The test failure looks unrelated

**Comment by geoand** [comment]
@kdubb what was the outcome of your experiment vs this?

**Comment by kdubb** [comment]
@tsegismont Is using the culmination of my original PR's against the vert.x sql client.

**Comment by geoand** [comment]
Excellent. @cescoffier are you still good with the latest version of the PR?

**Comment by cescoffier** [comment]
Yes, looks good to me.

**Comment by kdubb** [comment]
Is there any chance of this being back ported to 2.x? `idle-timeout` is the solution to many issues and discussions in the vault extension. It would be nice to have the functionality in both series if possible.

**Comment by geoand** [comment]
> Is there any chance of this being back ported to 2.x

-1 as we really don't want to backport features to 2.x

**Review by cescoffier — APPROVED** [review]

**Comment by geoand** [comment]
Can we get one last rebase please so we can make sure CI is fine?

**Comment by geoand** [comment]
Thanks!

**Comment by tsegismont** [comment]
@geoand this time it seems the build failed because one job ran out of time.

**Comment by geoand** [comment]
Yeah, I doubt it's related. @sberyozkin can you confirm?

**Comment by tsegismont** [comment]
@cescoffier @geoand can we go ahead and merge the PR?

**Comment by tsegismont** [comment]
Thanks @geoand !

**Comment by geoand** [comment]
Thank you!

**Pull request merged**
